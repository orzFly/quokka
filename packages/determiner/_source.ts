import type { RandomSource } from "./_generators.ts";

export function mathRandomSource(len: number): Uint8Array {
  const buffer = new Uint8Array(len);
  for (let i = 0; i < len; i++) {
    buffer[i] = Math.floor(Math.random() * 0x100);
  }
  return buffer;
}

export function webCryptoRandomSource(
  len: number,
  webCryptoObject?: Crypto,
): Uint8Array {
  const buffer = new Uint8Array(len);
  const target = webCryptoObject ?? crypto;
  target.getRandomValues(buffer);
  return buffer;
}

export function batchedRandomSource(
  source: RandomSource,
  batchSize: number,
): RandomSource {
  let buffer: Uint8Array | undefined;
  let offset = 0;

  return (len: number) => {
    if (!buffer) buffer = source(batchSize);

    const result = new Uint8Array(len);
    let pos = 0;
    while (pos < len) {
      const size = Math.min(len - pos, batchSize - offset);
      result.set(buffer.subarray(offset, offset + size), pos);
      pos += size;
      offset += size;
      if (offset === batchSize) {
        buffer = source(batchSize);
        offset = 0;
      }
    }
    return result;
  };
}
