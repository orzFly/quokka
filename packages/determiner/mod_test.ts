import { assert, assertEquals, assertExists } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { batchedRandomSource } from "./_source.ts";
import {
  bytes,
  Determiner,
  integer,
  pattern,
  uint16,
  uint32,
  uint64,
  uint8,
  uniquePatterns,
} from "./mod.ts";

const stableRandomSource = (len: number) => {
  const buffer = new Uint8Array(len);
  for (let i = 0; i < len; i++) {
    buffer[i] = i % 0x100;
  }
  return buffer;
};

const createSource = () => batchedRandomSource(stableRandomSource, 1024);

Deno.test("bytes function", async (t) => {
  const source = createSource();
  const result = bytes(source, 10);
  assertEquals(result.length, 10);
  assertEquals(result[0], 0);
  assertEquals(result[9], 9);
  await assertSnapshot(t, result);
});

Deno.test("uint64 function", async (t) => {
  const source = createSource();
  const result = uint64(source);
  assert(typeof result === "bigint");
  assert(result >= 0n);
  assert(result <= 0xFFFFFFFFFFFFFFFFn);
  await assertSnapshot(t, result);
});

Deno.test("uint32 function", async (t) => {
  const source = createSource();
  const result = uint32(source);
  assert(typeof result === "number");
  assert(result >= 0);
  assert(result <= 0xFFFFFFFF);
  await assertSnapshot(t, result);
});

Deno.test("uint16 function", async (t) => {
  const source = createSource();
  const result = uint16(source);
  assert(typeof result === "number");
  assert(result >= 0);
  assert(result <= 0xFFFF);
  await assertSnapshot(t, result);
});

Deno.test("uint8 function", async (t) => {
  const source = createSource();
  const result = uint8(source);
  assert(typeof result === "number");
  assert(result >= 0);
  assert(result <= 0xFF);
  await assertSnapshot(t, result);
});

Deno.test("integer function - default range", async (t) => {
  const source = createSource();
  const result = integer(source);
  assert(typeof result === "number");
  assert(result >= 0);
  assert(result <= 65536);
  await assertSnapshot(t, result);
});

Deno.test("integer function - custom range", async (t) => {
  const source = createSource();
  const result = integer(source, 5, 15);
  assert(typeof result === "number");
  assert(result >= 5);
  assert(result <= 15);
  await assertSnapshot(t, result);
});

Deno.test("integer function - invalid range", async (t) => {
  const source = createSource();
  const result = integer(source, 15, 5);
  assert(isNaN(result));
  await assertSnapshot(t, result);
});

Deno.test("pattern function - string pattern", async (t) => {
  const source = createSource();
  const result = pattern(source, 5, "abc");
  assertEquals(result.length, 5);
  assert(result.split("").every((char) => "abc".includes(char)));
  await assertSnapshot(t, result);
});

Deno.test("pattern function - array pattern", async (t) => {
  const source = createSource();
  const result = pattern(source, 3, ["abc", "123"]);
  assertEquals(result.length, 3);
  await assertSnapshot(t, result);
});

Deno.test("pattern function - default pattern", async (t) => {
  const source = createSource();
  const result = pattern(source, 8);
  assertEquals(result.length, 8);
  await assertSnapshot(t, result);
});

Deno.test("uniquePatterns function", async (t) => {
  const source = createSource();
  const result = uniquePatterns(source, 3, 4);
  assertEquals(result.length, 3);
  assertEquals(new Set(result).size, 3); // All patterns should be unique
  assert(result.every((pattern) => pattern.length === 4));
  await assertSnapshot(t, result);
});

Deno.test("uniquePatterns function with existing patterns", async (t) => {
  const source = createSource();
  const existing = ["abcd", "efgh"];
  const result = uniquePatterns(source, 2, 4, undefined, existing);
  assertEquals(result.length, 2);
  assert(!result.some((pattern) => existing.includes(pattern)));
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - constructor", () => {
  const source = createSource();
  const determiner = new Determiner(source);
  assertExists(determiner.source);
  assertEquals(determiner.maxUInt32, 4294967296);
});

Deno.test("Determiner class - bytes method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.bytes(5);
  assertEquals(result.length, 5);
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - uint64 method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.uint64();
  assert(typeof result === "bigint");
  assert(result >= 0n);
  assert(result <= 0xFFFFFFFFFFFFFFFFn);
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - uint32 method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.uint32();
  assert(typeof result === "number");
  assert(result >= 0);
  assert(result <= 0xFFFFFFFF);
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - uint16 method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.uint16();
  assert(typeof result === "number");
  assert(result >= 0);
  assert(result <= 0xFFFF);
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - uint8 method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.uint8();
  assert(typeof result === "number");
  assert(result >= 0);
  assert(result <= 0xFF);
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - integerLessThan method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.integerLessThan(100);
  assert(typeof result === "number");
  assert(result >= 0);
  assert(result < 100);
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - integer method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.integer(10, 20);
  assert(typeof result === "number");
  assert(result >= 10);
  assert(result <= 20);
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - pattern method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.pattern(6, "xyz");
  assertEquals(result.length, 6);
  assert(result.split("").every((char) => "xyz".includes(char)));
  await assertSnapshot(t, result);
});

Deno.test("Determiner class - uniquePatterns method", async (t) => {
  const source = createSource();
  const determiner = new Determiner(source);
  const result = determiner.uniquePatterns(4, 3);
  assertEquals(result.length, 4);
  assertEquals(new Set(result).size, 4);
  assert(result.every((pattern) => pattern.length === 3));
  await assertSnapshot(t, result);
});
