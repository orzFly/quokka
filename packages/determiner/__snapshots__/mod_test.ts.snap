export const snapshot = {};

snapshot[`bytes function 1`] = `
Uint8Array(10) [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
]
`;

snapshot[`uint64 function 1`] = `283686952306183n`;

snapshot[`uint32 function 1`] = `66051`;

snapshot[`uint16 function 1`] = `1`;

snapshot[`uint8 function 1`] = `0`;

snapshot[`integer function - default range 1`] = `514`;

snapshot[`integer function - custom range 1`] = `12`;

snapshot[`integer function - invalid range 1`] = `NaN`;

snapshot[`pattern function - string pattern 1`] = `"acbac"`;

snapshot[`pattern function - array pattern 1`] = `"b33"`;

snapshot[`pattern function - default pattern 1`] = `"b4vutsrq"`;

snapshot[`uniquePatterns function 1`] = `
[
  "b4kj",
  "a3ba",
  "t621",
]
`;

snapshot[`uniquePatterns function with existing patterns 1`] = `
[
  "a3ba",
  "o1dc",
]
`;

snapshot[`Determiner class - bytes method 1`] = `
Uint8Array(5) [
  0,
  1,
  2,
  3,
  4,
]
`;

snapshot[`Determiner class - uint64 method 1`] = `283686952306183n`;

snapshot[`Determiner class - uint32 method 1`] = `66051`;

snapshot[`Determiner class - uint16 method 1`] = `1`;

snapshot[`Determiner class - uint8 method 1`] = `0`;

snapshot[`Determiner class - integerLessThan method 1`] = `51`;

snapshot[`Determiner class - integer method 1`] = `17`;

snapshot[`Determiner class - pattern method 1`] = `"yxzyxz"`;

snapshot[`Determiner class - uniquePatterns method 1`] = `
[
  "b4h",
  "v81",
  "j2b",
  "p2v",
]
`;
